src/rgw/rgw_kafka.cc

```diff
@@ -79,40 +118,31 @@ struct connection_t {
   // cleanup of all internal connection resource
   // the object can still remain, and internal connection
   // resources created again on successful reconnection
-  void destroy(int s) {
-    status = s;
-    // destroy temporary conf (if connection was never established)
-    if (temp_conf) {
-        rd_kafka_conf_destroy(temp_conf);
-        return;
-    }
-    if (!is_ok()) {
+  void destroy() {
+    if (!producer) {
       // no producer, nothing to destroy
       return;
     }
-    // wait for all remaining acks/nacks
-    rd_kafka_flush(producer, 5*1000 /* wait for max 5 seconds */);
+    // wait for 5 seconds to try and handle pending callbacks
+    rd_kafka_flush(producer, 5*1000);
     // destroy all topics
     std::for_each(topics.begin(), topics.end(), [](auto topic) {rd_kafka_topic_destroy(topic);});
+    topics.clear();
     // destroy producer
     rd_kafka_destroy(producer);
     producer = nullptr;
     // fire all remaining callbacks (if not fired by rd_kafka_flush)
     std::for_each(callbacks.begin(), callbacks.end(), [this](auto& cb_tag) {
-        cb_tag.cb(status);
-        ldout(cct, 20) << "Kafka destroy: invoking callback with tag="
+        ldout(cct, 1) << "Kafka destroy: invoking callback with tag: "
                        << cb_tag.tag << " for: " << broker
-                       << " with status: " << status << dendl;
+                       << " with status: " << status_to_string(status) << dendl;
+        cb_tag.cb(status_to_errno(status));
```

Comment:
status default value = 0.
so this will evict the entries from the queue,
probably explicitly send `DESTROY` or some sensible message here ?

Reply:
right. destroy() is called either upon idleness (where state will be set accordingly).
or on shutdown (where state reflect last error or just zero).
will set state on shutdown

Reply:
changed status of all connections i nthe manager's dtor